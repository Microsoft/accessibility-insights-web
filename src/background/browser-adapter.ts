// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ClientBrowserAdapter, ClientChromeAdapter } from '../common/client-browser-adapter';

export interface NotificationOptions {
    message: string;
    title: string;
    iconUrl: string;
    type?: string;
}

export interface BrowserAdapter extends ClientBrowserAdapter {
    getAllWindows(getInfo: chrome.windows.GetInfo, callback: (chromeWindows: chrome.windows.Window[]) => void);
    getSelectedTabInWindow(windowId: number, callback: (activeTab: chrome.tabs.Tab) => void);
    addListenerToTabsOnActivated(callback: (activeInfo: chrome.tabs.TabActiveInfo) => void);
    addListenerToTabsOnUpdated(callback: (tabId: number, changeInfo: chrome.tabs.TabChangeInfo, tab: chrome.tabs.Tab) => void);
    addListenerToTabsOnRemoved(callback: (tabId: number, removeInfo: chrome.tabs.TabRemoveInfo) => void);
    addListenerToWebNavigationUpdated(callback: (details: chrome.webNavigation.WebNavigationFramedCallbackDetails) => void);
    addListenerOnWindowsFocusChanged(callback: (windowId: number) => void);
    tabsExecuteScript(tabId: number, details: chrome.tabs.InjectDetails, callback?: (result: any[]) => void);
    tabsQuery(query: chrome.tabs.QueryInfo, callback: (result: chrome.tabs.Tab[]) => void): void;
    createTab(url: string, callback?: (tab: chrome.tabs.Tab) => void);
    createTabInNewWindow(url: string, callback?: (tab: chrome.tabs.Tab) => void);
    createInactiveTab(url: string, callback: (tab: chrome.tabs.Tab) => void);
    closeTab(tabId: number);
    switchToTab(tabId: number);
    getTab(tabId: number, callback: (tab: chrome.tabs.Tab) => void);
    sendMessageToFramesAndTab(tabId: number, message: any): void;
    sendMessageToFrames(message: any);
    sendMessageToAllFramesAndTabs(message: any): void;
    setUserData?(items: Object, callback?: () => void);
    getUserData?(keys: string | string[] | Object, callback: (items: { [key: string]: any; }) => void);
    removeUserData?(key: string);
    injectJs(tabId, file: string, callback: Function);
    injectCss(tabId, file: string, callback: Function);
    getRunTimeId(): string;
    createNotification(options: NotificationOptions): void;
    getRuntimeLastError();
    isAllowedFileSchemeAccess?(callback: Function);
    addListenerToLocalStorage?(callback: (changes: object) => void);
    addCommandListener(callback: (command: string) => void): void;
    getCommands(callback: (commands: chrome.commands.Command[]) => void): void;
    createPopupWindow(url: string, callback: (window: chrome.windows.Window) => void): void;
}

export class ChromeAdapter extends ClientChromeAdapter implements BrowserAdapter {

    public getAllWindows(getInfo: chrome.windows.GetInfo, callback: (chromeWindows: chrome.windows.Window[]) => void): void {
        chrome.windows.getAll(getInfo, callback);
    }

    public getSelectedTabInWindow(windowId: number, callback: (activeTab: chrome.tabs.Tab) => void): void {
        chrome.tabs.getSelected(windowId, callback);
    }

    public addListenerToTabsOnActivated(callback: (activeInfo: chrome.tabs.TabActiveInfo) => void): void {
        chrome.tabs.onActivated.addListener(callback);
    }

    public addListenerToTabsOnUpdated(callback: (tabId: number, changeInfo: chrome.tabs.TabChangeInfo, tab: chrome.tabs.Tab) => void) {
        chrome.tabs.onUpdated.addListener(callback);
    }

    public addListenerToWebNavigationUpdated(callback: (details: chrome.webNavigation.WebNavigationFramedCallbackDetails) => void) {
        chrome.webNavigation.onDOMContentLoaded.addListener(callback);
    }

    public addListenerToTabsOnRemoved(callback: (tabId: number, removeInfo: chrome.tabs.TabRemoveInfo) => void) {
        chrome.tabs.onRemoved.addListener(callback);
    }

    public addListenerOnWindowsFocusChanged(callback: (windowId: number) => void) {
        chrome.windows.onFocusChanged.addListener(callback);
    }

    public getRunTimeId(): string {
        return chrome.runtime.id;
    }

    public tabsExecuteScript(tabId: number, details: chrome.tabs.InjectDetails, callback?: (result: any[]) => void) {
        chrome.tabs.executeScript(tabId, details, callback);
    }

    public tabsQuery(query: chrome.tabs.QueryInfo, callback: (result: chrome.tabs.Tab[]) => void): void {
        chrome.tabs.query(query, callback);
    }

    public getTab(tabId: number, callback: (tab: chrome.tabs.Tab) => void): void {
        chrome.tabs.get(tabId, callback);
    }

    public injectJs(tabId, file: string, callback?: (result: any[]) => void) {
        chrome.tabs.executeScript(tabId, {
            allFrames: true,
            file: file,
        }, callback);
    }
    public injectCss(tabId, file: string, callback?: Function) {
        chrome.tabs.insertCSS(tabId, {
            allFrames: true,
            file: file,
        }, callback);
    }

    public createTab(url: string, callback?: (tab: chrome.tabs.Tab) => void) {
        chrome.tabs.create({
            url: url,
            active: true,
            pinned: false,
        },
            callback,
        );
    }

    public createTabInNewWindow(url: string, callback?: (tab: chrome.tabs.Tab) => void) {
        chrome.windows.create({
            url: url,
            focused: true,
        }, window => {
                callback(window.tabs[0]);
            },
        );
    }

    public createPopupWindow(url: string, callback: (window: chrome.windows.Window) => void) {
        const height = 700;
        const width = 1000;
        const left = Math.floor(screen.width ? (screen.width - width) / 2 : 0);
        const top = Math.floor(screen.height ? (screen.height - height) / 2 : 0);

        chrome.windows.create({
            url: url,
            left: left,
            top: top,
            width: width,
            height: height,
            type: 'popup',
        }, window => {
            callback(window);
        });
    }

    public createInactiveTab(url: string, callback: (tab: chrome.tabs.Tab) => void) {
        chrome.tabs.create({
            url: url,
            active: false,
            pinned: false,
        },
            callback,
        );
    }

    public closeTab(tabId: number) {
        chrome.tabs.remove(tabId);
    }

    public switchToTab(tabId: number): void {
        const props = {
            active: true,
        };

        chrome.tabs.update(tabId, props, tab => {
            chrome.windows.update(tab.windowId, { focused: true });
        });
    }

    public sendMessageToFramesAndTab(tabId: number, message: any): void {
        chrome.runtime.sendMessage(message);
        chrome.tabs.sendMessage(tabId, message);
    }

    public sendMessageToAllFramesAndTabs(message: any): void {
        chrome.runtime.sendMessage(message);

        chrome.tabs.query({}, tabs => {
            for (let i = 0; i < tabs.length; ++i) {
                chrome.tabs.sendMessage(tabs[i].id, message);
            }
        });
    }

    public sendMessageToFrames(message: any) {
        chrome.runtime.sendMessage(message);
    }

    public setUserData(items: Object, callback?: () => void) {
        chrome.storage.local.set(items, callback);
    }

    public getUserData(keys: string | string[] | Object, callback: (items: { [key: string]: any; }) => void) {
        chrome.storage.local.get(keys, callback);
    }

    public removeUserData(key: string): void {
        chrome.storage.local.remove(key);
    }

    public getRuntimeLastError(): any {
        return chrome.runtime.lastError;
    }

    public createNotification(options: NotificationOptions): void {
        chrome.notifications.create({
            type: options.type || 'basic',
            iconUrl: options.iconUrl,
            title: options.title,
            message: options.message,
        });
    }

    public isAllowedFileSchemeAccess(callback: (isAllowed: boolean) => void): void {
        chrome.extension.isAllowedFileSchemeAccess(callback);
    }

    public addListenerToLocalStorage(callback: (changes: object) => void): void {
        chrome.storage.onChanged.addListener(callback);
    }

    public addCommandListener(callback: (command: string) => void): void {
        chrome.commands.onCommand.addListener(callback);
    }

    public getCommands(callback: (commands: chrome.commands.Command[]) => void): void {
        chrome.commands.getAll(callback);
    }
}
