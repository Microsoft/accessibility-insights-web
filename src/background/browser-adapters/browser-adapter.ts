// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ClientBrowserAdapter } from '../../common/client-browser-adapter';

export interface NotificationOptions {
    message: string;
    title: string;
    iconUrl: string;
    notificationType?: string;
}

export interface BrowserAdapter extends ClientBrowserAdapter {
    getAllWindows(getInfo: chrome.windows.GetInfo, callback: (chromeWindows: chrome.windows.Window[]) => void): void;
    getSelectedTabInWindow(windowId: number, callback: (activeTab: chrome.tabs.Tab) => void): void;
    addListenerToTabsOnActivated(callback: (activeInfo: chrome.tabs.TabActiveInfo) => void): void;
    addListenerToTabsOnUpdated(callback: (tabId: number, changeInfo: chrome.tabs.TabChangeInfo, tab: chrome.tabs.Tab) => void): void;
    addListenerToTabsOnRemoved(callback: (tabId: number, removeInfo: chrome.tabs.TabRemoveInfo) => void): void;
    addListenerToWebNavigationUpdated(callback: (details: chrome.webNavigation.WebNavigationFramedCallbackDetails) => void): void;
    addListenerOnWindowsFocusChanged(callback: (windowId: number) => void): void;
    tabsQuery(query: chrome.tabs.QueryInfo, callback: (result: chrome.tabs.Tab[]) => void): void;
    createTab(url: string, callback?: (tab: chrome.tabs.Tab) => void): void;
    createTabInNewWindow(url: string, callback?: (tab: chrome.tabs.Tab) => void): void;
    createInactiveTab(url: string, callback: (tab: chrome.tabs.Tab) => void): void;
    closeTab(tabId: number): void;
    switchToTab(tabId: number): void;
    getTab(tabId: number, onResolve: (tab: chrome.tabs.Tab) => void, onReject?: () => void): void;
    sendMessageToFramesAndTab(tabId: number, message: any): void;
    sendMessageToFrames(message: any): void;
    sendMessageToAllFramesAndTabs(message: any): void;
    injectJs(tabId, file: string, callback: Function): void;
    injectCss(tabId, file: string, callback: Function): void;
    getRunTimeId(): string;
    createNotification(options: NotificationOptions): void;
    getRuntimeLastError(): chrome.runtime.LastError;
    isAllowedFileSchemeAccess(callback: Function): void;
    addListenerToLocalStorage(callback: (changes: object) => void): void;
    openManageExtensionPage(): void;
}
