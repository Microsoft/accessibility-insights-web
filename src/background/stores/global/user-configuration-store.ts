// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { autobind } from '@uifabric/utilities';
import { cloneDeep } from 'lodash';

import { FeatureFlags } from '../../../common/feature-flags';
import { IndexedDBAPI } from '../../../common/indexedDB/indexedDB';
import { StoreNames } from '../../../common/stores/store-names';
import { UserConfigurationStoreData } from '../../../common/types/store-data/user-configuration-store';
import {
    SetBugServicePayload,
    SetBugServicePropertyPayload,
    SetHighContrastModePayload,
    SetIssueTrackerPathPayload,
    SetTelemetryStatePayload,
} from '../../actions/action-payloads';
import { FeatureFlagPayload } from '../../actions/feature-flag-actions';
import { UserConfigurationActions } from '../../actions/user-configuration-actions';
import { IndexedDBDataKeys } from '../../IndexedDBDataKeys';
import { BaseStore } from '../base-store';

export class UserConfigurationStore extends BaseStore<UserConfigurationStoreData> {
    public static readonly defaultState: UserConfigurationStoreData = {
        isFirstTime: true,
        enableTelemetry: false,
        enableHighContrast: false,
        bugService: 'none',
        bugServicePropertiesMap: {},
    };

    constructor(
        private readonly initialState: UserConfigurationStoreData,
        private readonly userConfigActions: UserConfigurationActions,
        private readonly indexDbApi: IndexedDBAPI,
    ) {
        super(StoreNames.UserConfigurationStore);
    }

    public getDefaultState(): UserConfigurationStoreData {
        return this.initialState ? cloneDeep(this.initialState) : cloneDeep(UserConfigurationStore.defaultState);
    }

    protected addActionListeners(): void {
        this.userConfigActions.getCurrentState.addListener(this.onGetCurrentState);
        this.userConfigActions.setTelemetryState.addListener(this.onSetTelemetryState);
        this.userConfigActions.setHighContrastMode.addListener(this.onSetHighContrastMode);
        this.userConfigActions.setBugService.addListener(this.onSetBugService);
        this.userConfigActions.setBugServiceProperty.addListener(this.onSetBugServiceProperty);
        this.userConfigActions.notifyFeatureFlagChange.addListener(this.onNotifyFeatureFlagChange);
        this.userConfigActions.setIssueTrackerPath.addListener(this.onSetIssueTrackerPath);
    }

    @autobind
    private onSetTelemetryState(payload: SetTelemetryStatePayload): void {
        this.state.isFirstTime = false;
        this.state.enableTelemetry = payload.enableTelemetry;

        // tslint:disable-next-line:no-floating-promises - grandfathered-in pre-existing violation
        this.indexDbApi.setItem(IndexedDBDataKeys.userConfiguration, this.state);
        this.emitChanged();
    }

    @autobind
    private onSetHighContrastMode(payload: SetHighContrastModePayload): void {
        this.state.enableHighContrast = payload.enableHighContrast;

        // tslint:disable-next-line:no-floating-promises - grandfathered-in pre-existing violation
        this.indexDbApi.setItem(IndexedDBDataKeys.userConfiguration, this.state);
        this.emitChanged();
    }

    @autobind
    private onSetBugService(payload: SetBugServicePayload): void {
        this.state.bugService = payload.bugServiceName;

        // tslint:disable-next-line:no-floating-promises - grandfathered-in pre-existing violation
        this.indexDbApi.setItem(IndexedDBDataKeys.userConfiguration, this.state);
        this.emitChanged();
    }

    @autobind
    private onSetBugServiceProperty(payload: SetBugServicePropertyPayload): void {
        if (!this.state.bugServicePropertiesMap) {
            this.state.bugServicePropertiesMap = {};
        }
        if (!this.state.bugServicePropertiesMap[payload.bugServiceName]) {
            this.state.bugServicePropertiesMap[payload.bugServiceName] = {};
        }

        this.state.bugServicePropertiesMap[payload.bugServiceName][payload.propertyName] = payload.propertyValue;

        // tslint:disable-next-line:no-floating-promises - grandfathered-in pre-existing violation
        this.indexDbApi.setItem(IndexedDBDataKeys.userConfiguration, this.state);
        this.emitChanged();
    }

    @autobind
    private onNotifyFeatureFlagChange(payload: FeatureFlagPayload): void {
        if (payload.feature === FeatureFlags.highContrastMode && payload.enabled === false) {
            this.onSetHighContrastMode({ enableHighContrast: false });
        }
    }

    @autobind
    private onSetIssueTrackerPath(payload: SetIssueTrackerPathPayload) {
        this.state.issueTrackerPath = payload.issueTrackerPath;

        // tslint:disable-next-line:no-floating-promises - grandfathered-in pre-existing violation
        this.indexDbApi.setItem(IndexedDBDataKeys.userConfiguration, this.state);
        this.emitChanged();
    }
}
