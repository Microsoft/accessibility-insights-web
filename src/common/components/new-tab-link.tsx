// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { BaseComponent } from '@uifabric/utilities';
import { ILinkProps, Link } from 'office-ui-fabric-react/lib/Link';
import * as React from 'react';

export class NewTabLink extends BaseComponent<ILinkProps> {
    public render(): JSX.Element {
        return <Link className="insights-link" {...this.props} target="_blank" />;
    }
}
