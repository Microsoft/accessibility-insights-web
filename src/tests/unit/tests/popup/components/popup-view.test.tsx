// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import * as React from 'react';
import { IMock, It, Mock, MockBehavior, Times } from 'typemoq';

import { BrowserAdapter } from '../../../../../common/browser-adapters/browser-adapter';
import { ChromeAdapter } from '../../../../../common/browser-adapters/chrome-adapter';
import { NewTabLink } from '../../../../../common/components/new-tab-link';
import { DropdownClickHandler } from '../../../../../common/dropdown-click-handler';
import { StoreActionMessageCreatorImpl } from '../../../../../common/message-creators/store-action-message-creator-impl';
import { BaseClientStoresHub } from '../../../../../common/stores/base-client-stores-hub';
import { LaunchPanelStoreData } from '../../../../../common/types/store-data/launch-panel-store-data';
import { UserConfigurationStoreData } from '../../../../../common/types/store-data/user-configuration-store';
import { PopupActionMessageCreator } from '../../../../../popup/actions/popup-action-message-creator';
import { LaunchPanelHeader } from '../../../../../popup/components/launch-panel-header';
import {
    LaunchPanelType,
    PopupView,
    PopupViewControllerDeps,
    PopupViewControllerState,
    PopupViewProps,
} from '../../../../../popup/components/popup-view';
import { DiagnosticViewClickHandler } from '../../../../../popup/handlers/diagnostic-view-toggle-click-handler';
import { PopupViewControllerHandler } from '../../../../../popup/handlers/popup-view-controller-handler';
import { LaunchPadRowConfigurationFactory } from '../../../../../popup/launch-pad-row-configuration-factory';
import { SupportLinkHandler } from '../../../../../popup/support-link-handler';
import { BaseDataBuilder } from '../../../common/base-data-builder';
import { IsSameObject } from '../../../common/typemoq-helper';

describe('PopupView', () => {
    const browserAdapterMock = Mock.ofType<BrowserAdapter>(ChromeAdapter);
    const launchPanelStateStoreState: LaunchPanelStoreData = {
        launchPanelType: LaunchPanelType.LaunchPad,
    };
    const featureFlagStoreData = {};

    afterEach(() => {
        browserAdapterMock.reset();
    });

    test('constructor', () => {
        const manifestStub = getManifestStub();
        const storesHubMock = createDefaultStoresHubMock(false);

        browserAdapterMock
            .setup(ba => ba.getManifest())
            .returns(() => manifestStub)
            .verifiable(Times.once());

        const props = createDefaultPropsBuilder(storesHubMock.object)
            .with('popupHandlers', {} as any)
            .build();

        (() => new PopupView(props))();

        storesHubMock.verifyAll();
    });

    test('render spinner', () => {
        const manifestStub = getManifestStub();

        const storesHubMock = createDefaultStoresHubMock(false, false);

        browserAdapterMock
            .setup(ba => ba.getManifest())
            .returns(() => manifestStub)
            .verifiable(Times.once());

        const props = createDefaultPropsBuilder(storesHubMock.object)
            .with('popupHandlers', {} as any)
            .with('hasAccess', true)
            .build();

        const testObject = new PopupView(props);

        expect(testObject.render()).toMatchSnapshot();
    });

    describe('render actual content', () => {
        let actionMessageCreatorStrictMock: IMock<PopupActionMessageCreator>;
        let dropdownClickHandlerMock: IMock<DropdownClickHandler>;
        let handlerMock: IMock<PopupViewControllerHandler>;
        let storesHubMock: IMock<BaseClientStoresHub<any>>;
        let clickHandlerMock: IMock<DiagnosticViewClickHandler>;
        let storeState: PopupViewControllerState;
        let deps: PopupViewControllerDeps;
        const manifestStub = getManifestStub();
        const rowConfigStub = {};
        const shortcutModifyHandlerStub = {};
        const launchPadRowConfigurationFactoryMock = Mock.ofType(LaunchPadRowConfigurationFactory);
        const popupViewStoreActionMessageCreatorMock = Mock.ofType(StoreActionMessageCreatorImpl);
        const userConfigStoreData: UserConfigurationStoreData = {
            isFirstTime: true,
            enableTelemetry: false,
            enableHighContrast: false,
            bugService: 'none',
            bugServicePropertiesMap: {},
        };

        beforeEach(() => {
            actionMessageCreatorStrictMock = Mock.ofType<PopupActionMessageCreator>(null, MockBehavior.Strict);
            dropdownClickHandlerMock = Mock.ofType(DropdownClickHandler);
            handlerMock = Mock.ofType(PopupViewControllerHandler);
            clickHandlerMock = Mock.ofType(DiagnosticViewClickHandler);
            storesHubMock = createDefaultStoresHubMock();
            browserAdapterMock
                .setup(ba => ba.getManifest())
                .returns(() => manifestStub)
                .verifiable();
            launchPadRowConfigurationFactoryMock
                .setup(l =>
                    l.createRowConfigs(It.isAny(), IsSameObject(actionMessageCreatorStrictMock.object), IsSameObject(handlerMock.object)),
                )
                .returns(() => rowConfigStub as any)
                .verifiable();

            storeState = {
                launchPanelStateStoreData: launchPanelStateStoreState,
                featureFlagStoreData: featureFlagStoreData,
                userConfigurationStoreData: userConfigStoreData,
            };

            deps = {
                popupActionMessageCreator: actionMessageCreatorStrictMock.object,
                dropdownClickHandler: dropdownClickHandlerMock.object,
                browserAdapter: browserAdapterMock.object,
            } as PopupViewControllerDeps;
        });

        test('render toggles view: launch pad', () => {
            actionMessageCreatorStrictMock.setup(acm => acm.openLaunchPad(launchPanelStateStoreState.launchPanelType)).verifiable();
            const props = createDefaultPropsBuilder(storesHubMock.object)
                .withDefaultTitleAndSubtitle()
                .with('deps', deps)
                .with('popupHandlers', {
                    diagnosticViewClickHandler: clickHandlerMock.object,
                    popupViewControllerHandler: handlerMock.object,
                    launchPanelHeaderClickHandler: null,
                    supportLinkHandler: new SupportLinkHandler(null, null),
                    shortcutModifyHandler: shortcutModifyHandlerStub as any,
                })
                .with('hasAccess', true)
                .with('launchPadRowConfigurationFactory', launchPadRowConfigurationFactoryMock.object)
                .with('storeState', storeState)
                .build();
            props.deps.storesHub = storesHubMock.object;
            props.deps.storeActionMessageCreator = popupViewStoreActionMessageCreatorMock.object;

            actionMessageCreatorStrictMock.setup(amc => amc.openTutorial(It.isAny()));

            const rendered = shallow(<PopupView {...props} />);

            expect(rendered.debug()).toMatchSnapshot();

            const Subtitle = () => rendered.find(LaunchPanelHeader).prop('subtitle') as JSX.Element;
            const renderedSubtitle = shallow(<Subtitle />);
            expect(renderedSubtitle.debug()).toMatchSnapshot('subtitle');
            const link = renderedSubtitle.find(NewTabLink);

            link.simulate('click');
            actionMessageCreatorStrictMock.verify(ac => ac.openTutorial(It.isAny()), Times.once());

            handlerMock.verifyAll();
        });

        test('render toggles view: ad-hoc tools', () => {
            const adHocLaunchPanelStateStoreState: LaunchPanelStoreData = {
                launchPanelType: LaunchPanelType.AdhocToolsPanel,
            };
            storeState.launchPanelStateStoreData = adHocLaunchPanelStateStoreState;
            actionMessageCreatorStrictMock.setup(acm => acm.openLaunchPad(adHocLaunchPanelStateStoreState.launchPanelType)).verifiable();
            const props = createDefaultPropsBuilder(storesHubMock.object)
                .withDefaultTitleAndSubtitle()
                .with('deps', deps)
                .with('popupHandlers', {
                    diagnosticViewClickHandler: clickHandlerMock.object,
                    popupViewControllerHandler: handlerMock.object,
                    launchPanelHeaderClickHandler: null,
                    supportLinkHandler: new SupportLinkHandler(null, null),
                    shortcutModifyHandler: shortcutModifyHandlerStub as any,
                })
                .with('hasAccess', true)
                .with('launchPadRowConfigurationFactory', launchPadRowConfigurationFactoryMock.object)
                .with('storeState', storeState)
                .build();
            props.deps.storesHub = storesHubMock.object;
            props.deps.storeActionMessageCreator = popupViewStoreActionMessageCreatorMock.object;

            const rendered = shallow(<PopupView {...props} />);

            expect(rendered.debug()).toMatchSnapshot();

            handlerMock.verifyAll();
        });

        test('renderAdHocToolsPanel', () => {
            const launchPanelStateStoreStateStub = {
                launchPanelType: LaunchPanelType.AdhocToolsPanel,
            };

            storeState.launchPanelStateStoreData = launchPanelStateStoreStateStub;

            actionMessageCreatorStrictMock.setup(acm => acm.openLaunchPad(launchPanelStateStoreStateStub.launchPanelType)).verifiable();

            const props = createDefaultPropsBuilder(storesHubMock.object)
                .withDefaultTitleAndSubtitle()
                .with('deps', deps)
                .with('popupHandlers', {
                    diagnosticViewClickHandler: null,
                    popupViewControllerHandler: handlerMock.object,
                    launchPanelHeaderClickHandler: null,
                    supportLinkHandler: new SupportLinkHandler(null, null),
                    shortcutModifyHandler: null,
                })
                .with('hasAccess', true)
                .with('diagnosticViewToggleFactory', null)
                .with('launchPadRowConfigurationFactory', launchPadRowConfigurationFactoryMock.object)
                .with('storeState', storeState)
                .build();
            props.deps.storesHub = storesHubMock.object;
            props.deps.storeActionMessageCreator = popupViewStoreActionMessageCreatorMock.object;
            const rendered = shallow(<PopupView {...props} />);

            expect(rendered.debug()).toMatchSnapshot();
            handlerMock.verifyAll();
        });
    });

    test('renderFailureMsgPanelForChromeUrl', () => {
        const manifestStub = getManifestStub();

        browserAdapterMock.setup(ba => ba.getManifest()).returns(() => manifestStub);

        const storesHubMock = createDefaultStoresHubMock();

        const props = createDefaultPropsBuilder(storesHubMock.object)
            .withDefaultTitleAndSubtitle()
            .with('popupHandlers', {
                gettingStartedDialogHandler: {} as any,
            } as any)
            .with('targetTabUrl', 'chrome://extensions')
            .with('hasAccess', false)
            .with('diagnosticViewToggleFactory', null)
            .build();

        const testObject = new PopupView(props);

        expect(testObject.render()).toMatchSnapshot();
    });

    test('renderFailureMsgPanelForFileUrl', () => {
        const manifestStub = getManifestStub();

        browserAdapterMock.setup(ba => ba.getManifest()).returns(() => manifestStub);
        const storesHubMock = createDefaultStoresHubMock();

        const props = createDefaultPropsBuilder(storesHubMock.object)
            .withDefaultTitleAndSubtitle()
            .with('popupHandlers', {
                gettingStartedDialogHandler: {} as any,
            } as any)
            .with('targetTabUrl', 'file:///')
            .with('hasAccess', false)
            .with('diagnosticViewToggleFactory', null)
            .build();

        const wrapped = shallow(<PopupView {...props} />);
        expect(wrapped.getElement()).toMatchSnapshot();
    });

    function createDefaultPropsBuilder(storeHub: BaseClientStoresHub<any>): PopupViewPropsBuilder {
        return new PopupViewPropsBuilder().withStoresHub(storeHub).withBrowserAdapter(browserAdapterMock.object);
    }

    function createDefaultStoresHubMock(hasStores = true, hasStoreData = true): IMock<BaseClientStoresHub<any>> {
        const storesHubMock = Mock.ofType(BaseClientStoresHub);
        storesHubMock.setup(s => s.hasStores()).returns(() => hasStores);
        storesHubMock.setup(s => s.hasStoreData()).returns(() => hasStoreData);
        storesHubMock.setup(s => s.addChangedListenerToAllStores(It.isAny()));
        storesHubMock.setup(s => s.removeChangedListenerFromAllStores(It.isAny()));
        return storesHubMock;
    }

    function getManifestStub(): chrome.runtime.Manifest {
        return {
            version: '2',
        } as chrome.runtime.Manifest;
    }
});

class PopupViewPropsBuilder extends BaseDataBuilder<PopupViewProps> {
    public withDefaultTitleAndSubtitle(): PopupViewPropsBuilder {
        this.data.title = 'test title';
        return this;
    }
    public withStoresHub(storesHub: BaseClientStoresHub<any>): PopupViewPropsBuilder {
        this.data.deps = {
            ...this.data.deps,
            storesHub,
        };

        return this;
    }

    public withBrowserAdapter(browserAdapter: BrowserAdapter): PopupViewPropsBuilder {
        this.data.deps = {
            ...this.data.deps,
            browserAdapter,
        };

        return this;
    }
}
