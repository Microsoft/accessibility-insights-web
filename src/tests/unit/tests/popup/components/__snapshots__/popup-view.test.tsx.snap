// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupView render actual content render toggles view: ad-hoc tools 1`] = `
"<Fragment>
  <div className=\\"ms-Fabric ad-hoc-tools-panel\\">
    <LaunchPanelHeader deps={{...}} title=\\"test title\\" supportLinkHandler={{...}} popupWindow={[undefined]} openAdhocToolsPanel={[Function]} featureFlags={{...}} />
    <AdHocToolsPanel backLinkHandler={[Function]} diagnosticViewToggleFactory={[undefined]} />
  </div>
  <TelemetryPermissionDialog deps={{...}} isFirstTime={true} />
</Fragment>"
`;

exports[`PopupView render actual content render toggles view: launch pad 1`] = `
"<Fragment>
  <div className=\\"ms-Fabric launch-panel\\" id=\\"new-launch-pad\\">
    <LaunchPanelHeader deps={{...}} title=\\"test title\\" subtitle={{...}} supportLinkHandler={{...}} popupWindow={[undefined]} openAdhocToolsPanel={[Function]} featureFlags={{...}} />
    <LaunchPad deps={{...}} productName=\\"test title\\" rowConfigs={[undefined]} version=\\"2\\" />
  </div>
  <TelemetryPermissionDialog deps={{...}} isFirstTime={true} />
</Fragment>"
`;

exports[`PopupView render actual content render toggles view: launch pad: subtitle 1`] = `
"<Fragment>
  <NewTabLink href=\\"https://go.microsoft.com/fwlink/?linkid=2082374\\" aria-label=\\"demo video\\" title=\\"watch the 3 minute video introduction\\" onClick={[Function: onClickTutorialLink]}>
    Watch 3-minute video introduction
  </NewTabLink>
   
</Fragment>"
`;

exports[`PopupView render actual content renderAdHocToolsPanel 1`] = `
"<Fragment>
  <div className=\\"ms-Fabric ad-hoc-tools-panel\\">
    <LaunchPanelHeader deps={{...}} title=\\"test title\\" supportLinkHandler={{...}} popupWindow={[undefined]} openAdhocToolsPanel={[Function]} featureFlags={{...}} />
    <AdHocToolsPanel backLinkHandler={[Function]} diagnosticViewToggleFactory={{...}} />
  </div>
  <TelemetryPermissionDialog deps={{...}} isFirstTime={true} />
</Fragment>"
`;

exports[`PopupView render spinner 1`] = `
<StyledSpinnerBase
  className="insights-spinner"
  label="Loading..."
  size={3}
/>
`;

exports[`PopupView renderFailureMsgPanelForChromeUrl 1`] = `
<div
  className="ms-Fabric unsupported-url-info-panel"
>
  <Header
    title="test title"
  />
  <div
    className="ms-Grid main-section"
  >
    <div
      className="launch-panel-general-container"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Accessibility Insights for Web can't run on this URL.</br>You'll need to go to a different page.",
        }
      }
    />
  </div>
</div>
`;

exports[`PopupView renderFailureMsgPanelForFileUrl 1`] = `
<FileUrlUnsupportedMessagePanel
  deps={
    Object {
      "browserAdapter": proxy {
        "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
        "addCommandListener": [Function],
        "addListenerOnConnect": [Function],
        "addListenerOnMessage": [Function],
        "addListenerOnWindowsFocusChanged": [Function],
        "addListenerToLocalStorage": [Function],
        "addListenerToTabsOnActivated": [Function],
        "addListenerToTabsOnRemoved": [Function],
        "addListenerToTabsOnUpdated": [Function],
        "addListenerToWebNavigationUpdated": [Function],
        "closeTab": [Function],
        "connect": [Function],
        "createInactiveTab": [Function],
        "createNotification": [Function],
        "createTab": [Function],
        "createTabInNewWindow": [Function],
        "executeScriptInTab": [Function],
        "extensionVersion": undefined,
        "getAllWindows": [Function],
        "getCommands": [Function],
        "getManageExtensionUrl": [Function],
        "getManifest": [Function],
        "getRunTimeId": [Function],
        "getRuntimeLastError": [Function],
        "getTab": [Function],
        "getUrl": [Function],
        "getUserData": [Function],
        "insertCSSInTab": [Function],
        "isAllowedFileSchemeAccess": [Function],
        "removeListenerOnMessage": [Function],
        "removeUserData": [Function],
        "sendMessageToAllFramesAndTabs": [Function],
        "sendMessageToFrames": [Function],
        "sendMessageToTab": [Function],
        "setUserData": [Function],
        "switchToTab": [Function],
        "tabsQuery": [Function],
      },
      "storesHub": proxy {
        "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
        "addChangedListenerToAllStores": [Function],
        "getAllStoreData": [Function],
        "hasStoreData": [Function],
        "hasStores": [Function],
        "removeChangedListenerFromAllStores": [Function],
        "stores": undefined,
      },
    }
  }
  header={
    <Header
      title="test title"
    />
  }
  title="test title"
/>
`;
